readthedocs/builds/tasks.py

```diff
@@ -0,0 +1,124 @@
+import logging
+
+from django.conf import settings
+from django.db.models import Avg
+
+from readthedocs.builds.models import Build, Version
+from readthedocs.projects.models import Feature
+
+log = logging.getLogger(__name__)
+
+
+class TaskRouter:
+
+    """
+    Celery tasks router.
+
+    It allows us to decide which queue is where we want to execute the task
+    based on project's settings but also in queue availability.
+
+    1. the project is using conda
+    2. new project with less than N successful builds
+    3. last N successful builds have a high time average
+
+    It ignores projects that have already set ``build_queue`` attribute.
+
+    https://docs.celeryproject.org/en/stable/userguide/routing.html#manual-routing
+    https://docs.celeryproject.org/en/stable/userguide/configuration.html#std:setting-task_routes
+    """
+
+    n_builds = 5
+    n_last_builds = 15
+    time_average = 350
+
+    BUILD_DEFAULT_QUEUE = 'build:default'
+    BUILD_LARGE_QUEUE = 'build:large'
+
+    def route_for_task(self, task, args, kwargs, **__):
+        log.info('Executing TaskRouter. task=%s', task)
+        if task not in (
+            'readthedocs.projects.tasks.update_docs_task',
+            'readthedocs.projects.tasks.sync_repository_task',
+        ):
+            log.info('Skipping routing non-build task. task=%s', task)
+            return
+
+        version = self._get_version(task, args, kwargs)
+        if not version:
+            log.info('No Build/Version found. No routing task. task=%s', task)
+            return settings.CELERY_DEFAULT_QUEUE
```

Comment:
Do we want this logic? Shouldn't we just return `None` and let Celery handle it?

Reply:
Yeah. `None` is better here. I'll update.

Reply:
Oh, wait. I got confused. I'm already returning `None` when it's not a Build task, but here, wouldn't be better to return `build:default`?

Reply:
I'm not sure -- we should just be consistent in the way we return when we aren't setting a specific queue. Otherwise I imagine it will confuse us in the future.